### PR TITLE
Fix issues with shorthand css properties

### PR DESCRIPTION
**Pull request by SylvainCorlay**
The `flex-flow` attribute is a shortcut containing the `flex-direction` and `flex-wrap`. 

When setting `flex-flow` to an empty string, this resets `flex-direction` to its default, that is `row`, even if the trait attribute for `flex-direction` was `column` which is a bug.

We actually don't need `flex-flow` which is only a shortcut to the two other properties so I am removing it.

### CODE REVIEW

**Comment by jdfreder** [comment]
I don't think we need to remove this trait.  Also, this heads us down the path where we are choosing a subset of the CSS layout arbitrarily.  How about checking the current value of the trait before setting it to `""`.  If the trait is undefined or null (w/e it is by default), don't reset to `""`.

**Comment by SylvainCorlay** [comment]
@jdfreder consider the following case. 

```Python
VBox([Button(), Button()])
```
The initial value for the trait `flex-direction` is then `'column'` and for `flex-flow` it is `''`. 

In the view constructor, we loop over these properties: We first set `style['flex-direction']` to `'column'`, which immediately changes the `flex-flow`attribute.
In the next iteration of the loop, we meet the `flex-flow` attribute. Setting it to `''`, we effectively replace the value of the `flex-direction` from `column` to `''`, that is `row nowrap`. I don't think that we can work around that in a simple way.

**Comment by SylvainCorlay** [comment]
For this reason, we cannot be stateful with css properties that are shorthand properties.

**Comment by jdfreder** [comment]
Ah yes, I see.  This is the only shorthand property?

**Comment by SylvainCorlay** [comment]
The only one that I noticed. There are others for which we don't expose the bound properties like `margin`.

**Comment by SylvainCorlay** [comment]
We need to be consistent about shorthand properties: either we only expose shorthands or we only expose bound properties - everywhere.

In the flexbox css guide, regarding the flex property, they recommend using the shorthand.

File: ipywidgets/static/widgets/js/layout.js [hunk 1]
```diff
@@ -104,7 +103,7 @@ define([
          */
         handleChange: function(trait, value) {
             this.displayed.then(_.bind(function(parent) {
-                if (parent && value) {
```
Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: Why did you change this behavior?

File: ipywidgets/widgets/widget_layout.py [hunk 2]
```diff
@@ -6,14 +6,14 @@
 from .widget import Widget, register
 from traitlets import Unicode, CUnicode
 
-@register('IPython.Button')
```
Comment (jdfreder) [inline ipywidgets/widgets/widget_layout.py]: You can see what file I used as a template ;)

File: ipywidgets/static/widgets/js/layout.js (hunk 1 above)
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: I am reverting a change I did in #285. Otherwise, one cannot set a property back to `''`.

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: Actually, I think this should be `(value !== undefined || value !== null)`

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: If you make this change, and make sure the traits default to `None`, I think you don't have to remove the shorthand trait

**Comment by SylvainCorlay** [comment]
@jdfreder this is ready for rereview.

File: ipywidgets/static/widgets/js/layout.js [hunk 3]
```diff
@@ -36,18 +36,15 @@ define([
                 'align_items',
                 'align_self',
                 'bottom',
+                'border',
```
Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: Do not add this, it's not layout

File: ipywidgets/widgets/widget_layout.py [hunk 4]
```diff
@@ -6,36 +6,41 @@
...
     https://developer.mozilla.org/en-US/docs/Web/CSS/Reference
+
+    When a property is also accessible via a shorthand property, we only
+    expose the shorthand.
+
+    For example:
+    - ``flex-grow``, ``flex-shrink`` and ``flex-basis`` are bound to ``flex``.
+    - ``flex-wrap`` and ``flex-direction`` are bound to ``flex-flow``.
+    - ``margin-[top/bottom/left/right]`` values are bound to ``margin``, etc.
     """
```
Comment (jdfreder) [inline ipywidgets/widgets/widget_layout.py]: Nice, I like this doc string :)

File: ipywidgets/widgets/widget_layout.py [hunk 5]
```diff
@@ -6,36 +6,41 @@
...
-    
+
     _view_name = Unicode('LayoutView', sync=True)
 
     # Keys
     align_content = CUnicode(sync=True, allow_none=True)
     align_items = CUnicode(sync=True, allow_none=True)
     align_self = CUnicode(sync=True, allow_none=True)
+    border = CUnicode(sync=True, allow_none=True)
```
Comment (jdfreder) [inline ipywidgets/widgets/widget_layout.py]: -1

File: ipywidgets/static/widgets/js/layout.js (hunk 3 above)
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: Well `border-width-[top/bottom/left/right]` is...

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: That's like arguing font size is layout because it can affect the width and height of the object depending on the contents.  I would not add this, it's primary function is to set border style.  Nor would I add border-width-top/bottom/left/right.  All of that stuff should be set by CSS and not accessible in `Layout`

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: I included it because it is part of the [css box model](http://www.w3.org/TR/CSS21/images/boxdim.png).

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: Just because the box model accounts for it in its calculations, doesn't mean it should be used to conduct layout.  Instead padding or margin should be used.

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: The nice demo notebook uses it for the wrapping example :(

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: Besides, I like the idea of being able to group items together in a widget dashboard using a border, like this.

sliders
)

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: And I like the idea of being able to set background color, but that's not Layout. ;)

But seriously, you can achieve the same affect fairly easily by using CSS and `add_class`.

```
%%html
<style>.bordered { border: 2px solid grey; }</style>
```

and

```python
orient.add_class('bordered')
speeds.add_class('bordered')
```

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: Yeah but it is not a python API anymore. It requires a magic...

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: I don't think we'll agree on this- we should probably try to get a second opinion: @lbustelo , @ellisonbg , or @jasongrout  , what are your opinions on adding `border` to the mutable `Layout` object?

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: @all it is not only about adding border, but about only using shorthand css properties when there is one in order to avoid the issues discussed in this issue. border is the top-most shorthand property covering the border widths, which concern layout in my opinion.

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: Here's an idea, just use `border-width`, since it doesn't conflate the styling with the layout, and it itselft is a top-level shorthand: `https://developer.mozilla.org/en-US/docs/Web/CSS/border-width

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: It was my initial idea, but the `border-width` is only used if `border-style` is not `none`. That is why I ended up using the `border` shorthand.

Comment (jdfreder) [inline ipywidgets/static/widgets/js/layout.js]: > It was my initial idea, but the border-width is only used if border-style is not none. That is why I ended up using the border shorthand.

So in order to use the layout part of border, you also need to set the styling part of border?  Now I definitely don't think border or border-width should be included.

Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/layout.js]: > > It was my initial idea, but the border-width is only used if border-style is not none. That is why I ended up using the border shorthand.

> So in order to use the layout part of border, you also need to set the styling part of border? 

That is right.

**Comment by SylvainCorlay** [comment]
After one on one discussion with @jdfreder I removed the `border` attribute so that we can merge the bug fix and discuss this later.

**Comment by jdfreder** [comment]
Thanks! :)

**Comment by jdfreder** [comment]
Test failure looks unrelated, and actually a problem with karma.

**Comment by SylvainCorlay** [comment]
Thanks, I was looking into the failures.

**Comment by ellisonbg** [comment]
The case of border width is a subtle one. If it is part of the box model,
then we may want to include it in layout.

On Tue, Jan 5, 2016 at 4:35 PM, Sylvain Corlay <notifications@github.com>
wrote:

> Thanks, I was looking into the failures.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/pull/286#issuecomment-169178709>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
